src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -71,92 +71,148 @@ private class ConfigurationCacheItem
+            };
 
-            FeatureDefinition featureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false);
+            await EvaluateFeature(evaluationEvent, appContext, useAppContext, cancellationToken);
 
-            VariantDefinition variantDefinition = null;
+            return evaluationEvent.IsEnabled;
+        }
 
-            if (featureDefinition != null)
+        public ValueTask<Variant> GetVariantAsync(string feature, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(feature))
             {
-                isFeatureEnabled = await IsEnabledAsync(featureDefinition, appContext, useAppContext, cancellationToken).ConfigureAwait(false);
+                throw new ArgumentNullException(nameof(feature));
+            }
+
+            return GetVariantEvaluation(feature, context: null, useContext: false, cancellationToken);
+        }
+
+        public ValueTask<Variant> GetVariantAsync(string feature, TargetingContext context, CancellationToken cancellationToken)
+        {
+            if (string.IsNullOrEmpty(feature))
+            {
+                throw new ArgumentNullException(nameof(feature));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            return GetVariantEvaluation(feature, context, useContext: true, cancellationToken);
+        }
 
-                if (featureDefinition.Variants != null && featureDefinition.Variants.Any() && featureDefinition.Allocation != null)
+        private async ValueTask<Variant> GetVariantEvaluation(string feature, TargetingContext context, bool useContext, CancellationToken cancellationToken)
+        {
+            EvaluationEvent evaluationEvent = new EvaluationEvent
+            {
+                FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
+            };
+
+            await EvaluateFeature(evaluationEvent, context, useContext, cancellationToken);
+
+            return evaluationEvent.Variant;
+        }
+
+        private async Task<EvaluationEvent> EvaluateFeature<TContext>(EvaluationEvent evaluationEvent, TContext context, bool useContext, CancellationToken cancellationToken)
+        {
+            if (evaluationEvent.FeatureDefinition != null)
+            {
+                //
+                // Determine IsEnabled
+                evaluationEvent.IsEnabled = await IsEnabledAsync(evaluationEvent.FeatureDefinition, context, useContext, cancellationToken).ConfigureAwait(false);
+
+                //
+                // Determine Variant
+                VariantDefinition variantDefinition;
+
+                if (evaluationEvent.FeatureDefinition.Allocation == null || (!evaluationEvent.FeatureDefinition.Variants?.Any() ?? false))
+                {
+                    variantDefinition = null;
+
+                    evaluationEvent.VariantReason = "No Allocation or Variants";
```

Comment:
Not sure if this is just unclear to me, but the message sounds like the feature is missing both allocation and variants

Reply:
Agreed, and it's supposed to be either. Any suggestions for a better one? Perhaps "Missing Allocation or Variants"...?

Reply:
Yea, I like "Missing Allocation or Variants" more.